input-events/input-events-typing.html

```diff
@@ -1,82 +1,311 @@
-                }
-                ++eventCounter;
-                if (eventCounter === expectedResult.length)
-                    t.done();
-            }));
+let inputEventsLog;
+
+function log(event) {
+    inputEventsLog.push({
+        type: event.type,
+        inputType: event.inputType,
+        data: event.data,
+        isComposing: event.isComposing,
+        cancelable: event.cancelable,
+        state: encodeURI(document.getElementById('rich').innerHTML)
+    });
+}
+
+function resetEditor(content = '') {
+    inputEventsLog = [];
+    document.getElementById('rich').innerHTML = content;
+}
+
+document.getElementById('rich').addEventListener('beforeinput', event => log(event));
+document.getElementById('rich').addEventListener('input', event => log(event));
+</script>
+<script>
+(function() {
+    promise_test(async () => {
+        resetEditor();
+        document.getElementById('rich').focus();
+        const message = 'Hello';
+        await test_driver.send_keys(document.getElementById('rich'), message);
+        // 10 events (5 beforeinput + 5 input events)
+        assert_equals(inputEventsLog.length, 10);
+        for (let i = 0; i < inputEventsLog.length; i += 2) {
+            const beforeInputEvent = inputEventsLog[i];
+            const inputEvent = inputEventsLog[i + 1];
+            assert_equals(beforeInputEvent.type, 'beforeinput');
+            assert_equals(inputEvent.type, 'input');
+            assert_equals(beforeInputEvent.inputType, 'insertText');
+            assert_equals(inputEvent.inputType, 'insertText');
+            assert_equals(beforeInputEvent.data, inputEvent.data);
+            assert_equals(inputEvent.data, message[i / 2]);
+            assert_equals(beforeInputEvent.state + message[i / 2], inputEvent.state);
+        }
+        resetEditor();
+        return Promise.resolve();
+    }, 'It triggers beforeinput and input events on text typing');
+
+    promise_test(async () => {
+        resetEditor();
+        document.getElementById('rich').focus();
+        await test_driver.send_keys(document.getElementById('rich'), "\uE006"); // Return
+
+        assert_equals(inputEventsLog.length, 2);
+        const beforeInputEvent = inputEventsLog[0];
+        const inputEvent = inputEventsLog[1];
+        assert_equals(beforeInputEvent.type, 'beforeinput');
+        assert_equals(inputEvent.type, 'input');
+        assert_equals(beforeInputEvent.inputType, 'insertParagraph');
+        assert_equals(inputEvent.inputType, 'insertParagraph');
+        assert_equals(beforeInputEvent.data, inputEvent.data);
+        resetEditor();
+        return Promise.resolve();
+    }, 'It triggers beforeinput and input events on typing RETURN');
+
+    promise_test(async () => {
+        resetEditor();
+        document.getElementById('rich').focus();
+        await new test_driver.Actions()
+            .keyDown('\uE008') // Shift
+            .keyDown('\uE006') // Enter
```

Comment:
nit: above comment says `// Return` ... maybe use the unicode standard says the name is?

Reply:
Done. I used the names on this list https://w3c.github.io/webdriver/#keyboard-actions